doc/demo/solvers.py

```diff
@@ -63,3 +72,152 @@ def __del__(self):
         self.solver.destroy()
         self.A.destroy()
         self.b.destroy()
+
+
+class NewtonProblem(NonlinearProblem):
+    """Problem for the DOLFINx NewtonSolver with an external callback."""
+
+    def __init__(
+        self,
+        F: ufl.form.Form,
+        u: fem.function.Function,
+        bcs: list[fem.bcs.DirichletBC] = [],
+        J: ufl.form.Form = None,
+        form_compiler_options: Optional[dict] = None,
+        jit_options: Optional[dict] = None,
+        external_callback: Optional[Callable] = None,
+    ):
+        super().__init__(F, u, bcs, J, form_compiler_options, jit_options)
+
+        self.external_callback = external_callback
+
+    def form(self, x: PETSc.Vec) -> None:
+        """This function is called before the residual or Jacobian is
+        computed. This is usually used to update ghost values, but here
+        we also use it to evaluate the external operators.
+
+        Args:
+           x: The vector containing the latest solution
+        """
+        # The following line is from the standard NonlinearProblem class
+        x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+
+        # The external operators are evaluated here
+        self.external_callback()
+
+
+class SNESProblem:
+    """Solves a nonlinear problem via PETSc.SNES.
+    F(u) = 0
+    J = dF/du
+    b = assemble(F)
+    A = assemble(J)
+    Ax = b
+    """
+
+    def __init__(
+        self,
+        u: fem.function.Function,
+        F: ufl.form.Form,
+        J: ufl.form.Form,
+        bcs: list[fem.bcs.DirichletBC] = [],
+        petsc_options: Optional[dict] = {},
+        prefix: Optional[str] = None,
+        external_callback: Optional[Callable] = None,
+    ):
+        self.u = u
+        V = self.u.function_space
+        self.comm = V.mesh.comm
+
+        self.F_form = fem.form(F)
+        # J = ufl.derivative(F_form, self.u, ufl.TrialFunction(V))
+        self.J_form = fem.form(J)
+        self.b = fem.petsc.create_vector(self.F_form)
+        self.A = fem.petsc.create_matrix(self.J_form)
+
+        self.bcs = bcs
+
+        # Give PETSc solver options a unique prefix
+        if prefix is None:
+            prefix = f"snes_{str(id(self))[0:4]}"
+
+        self.prefix = prefix
+        self.petsc_options = petsc_options
+        self.external_callback = external_callback
+
+        self.solver = self.solver_setup()
+
+    def set_petsc_options(self):
```

Comment:
Not necessary if you let the user set up the SNES object via the first-class API.

Reply:
If the user passes petsc options, but never calls this method, the class is in an incomplete state.